policysync/processor_test.go

```diff
@@ -293,6 +298,314 @@ var _ = Describe("Processor", func() {
 
 			})
 		})
+
+		Describe("IP Set updates", func() {
+
+			Context("with two joined endpoints, one with active profile", func() {
+				var refdOutput chan proto.ToDataplane
+				var unrefdOutput chan proto.ToDataplane
+				var refdId proto.WorkloadEndpointID
+				var unrefdId proto.WorkloadEndpointID
+				var assertInactiveNoUpdate func()
+
+				BeforeEach(func(done Done) {
+					refdId = testId("refd")
+					refdOutput, _ = join("refd")
+					unrefdId = testId("unrefd")
+					unrefdOutput, _ = join("unrefd")
+
+					// Ensure the joins are completed by sending a workload endpoint for each.
+					updates <- &proto.WorkloadEndpointUpdate{
+						Id:       &refdId,
+						Endpoint: &proto.WorkloadEndpoint{},
+					}
+					<-refdOutput
+					updates <- &proto.WorkloadEndpointUpdate{
+						Id:       &unrefdId,
+						Endpoint: &proto.WorkloadEndpoint{},
+					}
+					<-unrefdOutput
+
+					// Send the IPSet, a Profile referring to it, and a WEP update referring to the
+					// Profile. This "activates" the WEP relative to the IPSet
+					updates <- updateIpSet(IPSetName, 0)
+					updates <- &proto.ActiveProfileUpdate{
+						Id: &proto.ProfileID{Name: ProfileName},
+						Profile: &proto.Profile{InboundRules: []*proto.Rule{
+							{
+								Action:      "allow",
+								SrcIpSetIds: []string{IPSetName},
+							},
+						}},
+					}
+					updates <- &proto.WorkloadEndpointUpdate{
+						Id:       &refdId,
+						Endpoint: &proto.WorkloadEndpoint{ProfileIds: []string{ProfileName}},
+					}
+					// All three updates get pushed to the active endpoint (1)
+					<-refdOutput
+					<-refdOutput
+					<-refdOutput
+
+					assertInactiveNoUpdate = func() {
+						// Send a WEP update for the inactive and check we get it from the output
+						// channel. This ensures that the inactive endpoint didn't get the IPSetUpdate
+						// without having to wait for a timeout.
+						updates <- &proto.WorkloadEndpointUpdate{
+							Id:       &unrefdId,
+							Endpoint: &proto.WorkloadEndpoint{},
+						}
+						wep := <-unrefdOutput
+						Expect(wep.GetWorkloadEndpointUpdate().GetId()).To(Equal(&unrefdId))
+					}
+
+					close(done)
+				})
+
+				It("should send IPSetUpdate to only to ref'd endpoint", func(done Done) {
+					msg := updateIpSet(IPSetName, 2)
+					updates <- msg
+					g := <-refdOutput
+					Expect(g).To(Equal(proto.ToDataplane{Payload: &proto.ToDataplane_IpsetUpdate{msg}}))
+
+					assertInactiveNoUpdate()
+					close(done)
```

Comment:
closing the `Done` channel just indicates to Ginkgo that the testcase is over. It's not a test cleanup operation.

Reply:
Ah, I didn't spot where that came from; I'm so used to seeing `It` lambdas with no arguments.  Why do you need the done argument here?

Reply:
I find the `done` construction easier to read than using `Eventually` everywhere.  It also gives more flexibility for matching specific fields of the channel message; `Eventually` only matches on the top level message coming out of the channel.

Reply:
OK, so the only purpose of adding the Done argument is to trigger Ginkgo's timeout logic.  Can't say I'm keen on their API choice; there's nothing about adding a Done argument to a test that says "now this will run with a timeout" to me.  It's unfortunate that they don't give you a way to shut down a test after it times out.  Presumably it'll be leaked.